test/runtests.jl

```diff
@@ -218,20 +293,31 @@ end
     end
 
     @testset "fft" begin
-        for x in (randn(3), randn(3, 4), randn(3, 4, 5))
+        for x in (randn(2), randn(2, 3), randn(3, 4, 5))
             N = ndims(x)
             complex_x = complex.(x)
             for dims in unique((1, 1:N, N))
+                # fft, ifft, bfft
                 for f in (fft, ifft, bfft)
                     test_frule(f, x, dims)
                     test_rrule(f, x, dims)
                     test_frule(f, complex_x, dims)
                     test_rrule(f, complex_x, dims)
                 end
+                for pf in (plan_fft, plan_ifft, plan_bfft) 
+                    test_frule(*, pf(x, dims) ⊢ NoTangent(), x)
```

Comment:
Why are the `NoTangent` needed here?

Reply:
It tells `ChainRulesTestUtils` to use a no tangent for the plan, i.e. do not try to differentiate w.r.t. the plan. This is necessary because with all the caching stuff and plans being mutable, ChainRules gets confused about the structure of `Plan`.

Unfortunately, there is one case where we do want to differentiate w.r.t. plan (as far as I can tell this is the only case), when someone makes a `ScaledPlan` whose scale depends on the parameter:

```
using AbstractFFTs
using AbstractFFTs: Plan
using Zygote
include("test/testplans.jl")

julia> function f(x)
           return sum(abs.(P * x))
           end
f (generic function with 1 method)

# correct 
julia> Zygote.gradient(f, [1,2,3])
([-0.732050807568877, 0.9999999999999996, 2.7320508075688776],)

julia> function f(x)
           return sum(abs.(x[1] * P * x))
           end
f (generic function with 1 method)

# silently wrong :(
julia> Zygote.gradient(f, [1,2,3])
([-0.732050807568877, 0.9999999999999996, 2.7320508075688776],)

```
I just spent some time trying to write an adjoint for `ScaledPlan` by replacing it with the right-associative `P.scale * (P.p * x)` and using `rrule_via_ad`, but the fundamental issue was that `ChainRules` was unable to come up with a tangent type for `ScaledPlan`, for the same reason (mutability, circular references, etc.)

A few thoughts:

- Really, even if we keep the `pinv` field around, `ScaledPlan` ought to not be mutable as the caching can just happen at the level of the inner plan. If this were fixed, it would be easy to write an `rrule`, but this would be a separate PR that would probably require a lot of careful thought
- Maybe we can somehow get the `ScaledPlan` differentiation to work, by adding a custom tangent type for the mutable struct. Don't have much ChainRules knowhow but I can give this a shot
- I really really don't like that this silently gives incorrect results (although the current plan rules in Zygote do too)

Reply:
Maybe making `ScaledPlan` immutable isn't so hard... give me a minute:)

Reply:
To try to resolve this, I've:

a) Made `ScaledPlan`s immutable in https://github.com/JuliaMath/AbstractFFTs.jl/pull/72 (and done the same for `AdjointPlan`s here

b) Added a backwards rule for `ScaledPlan that differentiates w.r.t. the plan's scale too

c) Added tests. Because of https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/256, this turned out to be rather difficult. I couldn't come up with a way of using `ChainRulesTestUtils` for properly checking the derivative w.r.t. `ScaledPlan` without a PR there, and I have to pick my battles, so I ended up coming up with the best cludge I can think to preserve most of the automated testing w.r.t. the other tangents and adding a manual FD test for the plan scale.